pkg/dns/dns_test.go

```diff
@@ -0,0 +1,282 @@
+}
+
+func (s *DNSTestSuite) Test_UDPProxy() {
+	mockedDNSConsulClient := pbdns.NewMockDNSServiceClient(s.T())
+	addr := &net.UDPAddr{IP: net.IPv4(127, 0, 0, 1), Port: 0}
+	connUdp, err := net.ListenUDP("udp", addr)
+	s.Require().NoError(err)
+	stopChan := make(chan struct{})
+	defer func() { stopChan <- struct{}{} }()
+
+	server := DNSServer{
+		client:  mockedDNSConsulClient,
+		connUDP: connUdp,
+		logger:  hclog.Default(),
+		stopCh:  stopChan,
+	}
+
+	go server.proxyUDP()
+
+	testCases := map[string]struct {
+		dnsRequest   []byte
+		dnsResp      []byte
+		expected     error
+		largeResp    error
+		expectedGRPC error
+	}{
+
+		"happy path": {
+			dnsRequest: genRandomBytes(512),
+			dnsResp:    genRandomBytes(50),
+		},
+		"happy large response path": {
+			dnsRequest: genRandomBytes(50),
+			dnsResp:    genRandomBytes(9216), // net.inet.udp.maxdgram for macs
+		},
+		"bad consul response too large": {
+			dnsRequest: genRandomBytes(50),
+			dnsResp:    genRandomBytes(65536),
+			expected:   errors.New("timeout"),
+		},
+		"bad consul response": {
+			dnsRequest:   genRandomBytes(512),
+			dnsResp:      genRandomBytes(50),
+			expectedGRPC: errors.New("timeout"),
+		},
+	}
+
+	for name, tc := range testCases {
+		s.Run(name, func() {
+
+			req := tc.dnsRequest
+			resp := tc.dnsResp
+
+			clientResp := &pbdns.QueryResponse{
+				Msg: resp,
+			}
+
+			mockedDNSConsulClient.On("Query", mock.Anything, mock.Anything).
+				Return(clientResp, tc.expectedGRPC).Once()
+			addr := fmt.Sprintf("127.0.0.1:%v", server.UdpPort())
+
+			conn, err := net.Dial("udp", addr)
+
+			s.Require().NoError(err)
+
+			n, err := conn.Write(req)
+			if err != nil {
+				s.T().Logf("error: %v", err.Error())
+			}
+			s.T().Logf("written %v", n)
+			p := make([]byte, 9216)
+			_ = conn.SetReadDeadline(time.Now().Add(time.Second * 1))
+			lengthRead, err := conn.Read(p)
+			s.T().Logf("read %v", lengthRead)
+			if tc.expectedGRPC != nil {
+				s.Require().Error(err)
+				s.Require().ErrorContains(err, tc.expectedGRPC.Error())
+			} else if tc.expected != nil {
+				s.Require().Error(err)
+				s.Require().ErrorContains(err, tc.expected.Error())
+				return
+			} else {
+				s.Require().NoError(err, "exchange error")
+				s.Require().EqualValues(resp, p[0:lengthRead])
+				s.Require().Equal(lengthRead, len(resp))
+			}
+			conn.Close()
+		})
+	}
+
+}
+
+func (s *DNSTestSuite) Test_ProxydnsTCP() {
+	mockedDNSConsulClient := pbdns.NewMockDNSServiceClient(s.T())
+	addr := &net.TCPAddr{IP: net.IPv4(127, 0, 0, 1), Port: 0}
+	listenerTCP, err := net.ListenTCP("tcp", addr)
+	s.Require().NoError(err)
+
+	stopChan := make(chan struct{})
+	defer close(stopChan)
+	server := DNSServer{
+		client:      mockedDNSConsulClient,
+		listenerTCP: listenerTCP,
+		logger:      hclog.Default(),
+		stopCh:      stopChan,
+	}
+
+	go server.proxyTCP()
+
+	testCases := map[string]struct {
+		dnsRequest   []byte
+		dnsResp      []byte
+		expected     error
+		largeResp    error
```

Comment:
nit:

```suggestion
		largeResp    bool
```

The error value appears to never be used but instead we only ever care whether its nil or not. In that case you could have this be a bool instead. You would also need to change the two places its referenced.